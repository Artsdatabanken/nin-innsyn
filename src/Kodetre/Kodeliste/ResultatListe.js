import { Divider, List, ListItem, ListItemText, Paper } from '@material-ui/core'
import React, { Component } from 'react'
import Backend from '../../backend'
import språk from '../../språk'

class ResultatListe extends Component {
  filtrer(kode) {
    const prefix = kode.substring(0, 2)
    switch (prefix) {
      case 'AR':
      case 'AO':
      case 'VV':
        return ''
      default:
        return kode.substring(3)
    }
  }

  render() {
    const { onClick, query, searchResults } = this.props
    if (!searchResults) return null
    if (!searchResults.length > 0) return null
    return (
      <Paper zDepth={1}>
        <List
          style={{
            overflow: 'hidden',
            maxHeight: 494,
            paddingTop: 0,
            paddingBottom: 0,
          }}
        >
          {searchResults.map(item => {
            const kode = item.kode.toUpperCase()
            const prefix = kode.substring(0, 2)
            const navn = språk(item.navn)

            return (
              <React.Fragment key={item.kode}>
                <ListItem
                  style={{
                    width: 392,
                    height: 38,
                    pointer: 'hand',
                    paddingTop: 0,
                    paddingBottom: 1,
                  }}
                  innerDivStyle={{
                    paddingTop: 14,
                    paddingLeft: 66,
                    fontSize: 13,
                    lineheight: 24,
                    fontWeight: 500,
                    whiteSpace: 'nowrap',
                    textOverflow: 'ellipsis',
                    overflow: 'hidden',
                  }}
                  onClick={() => {
                    onClick(item.kode)
                  }}
                  key={kode}
                  leftIcon={
                    <img
                      style={{ marginTop: 6, marginLeft: 16 }}
                      alt={prefix}
                      src={Backend.avatar24px(prefix)}
                      onError={e => {
                        const brokenImage = Backend.avatar24px('404')
                        if (e.target.src !== brokenImage)
                          e.target.src = brokenImage
                      }}
                    />
                  }
                  rightAvatar={
                    <div
                      style={{
                        paddingTop: 6,
                        fontWeight: 800,
                      }}
                    >
                      {ResultatListe.highlightMatch(this.filtrer(kode), query)}
                    </div>
                  }
                >
                  <ListItemText>
                    {ResultatListe.highlightMatch(navn, query)}
                  </ListItemText>
                </ListItem>
                <Divider inset={true} />
              </React.Fragment>
            )
          })}
        </List>
      </Paper>
    )
  }

  // Highlight all matches
  static highlightMatch(text, higlight) {
    // make array of terms, ordered by longest term
    let terms = higlight
      .toLowerCase()
      .split(' ')
      .sort(function(a, b) {
        return b.length - a.length
      })
    // make regex OR filter by concatenating terms with |
    let filter = terms
      .toString()
      .toLowerCase()
      .replace(/[,\\]/g, '|')

    // Split on all terms and also include the terms into parts array, ignore case
    let parts = text.split(new RegExp(`(${filter})`, 'gi'))
    return (
      <React.Fragment>
        {' '}
        {parts.map((part, i) => (
          <span
            key={i}
            style={
              terms.indexOf(part.toLowerCase()) >= 0
                ? { color: 'black', fontWeight: 'bold' }
                : {}
            }
          >
            {part}
          </span>
        ))}{' '}
      </React.Fragment>
    )
  }
}

export default ResultatListe
