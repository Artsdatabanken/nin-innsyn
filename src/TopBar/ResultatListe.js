import { Divider, ListItem, ListItemText } from "@material-ui/core";
import { withStyles } from "@material-ui/core/styles";
import React, { Component } from "react";
import Bildeavatar from "GjenbruksElement/Bildeavatar";

const styles = {};

class ResultatListe extends Component {
  filtrer(kode) {
    const prefix = kode.substring(0, 2);
    switch (prefix) {
      case "AR":
      case "AO":
      case "VV":
        return "";
      default:
        return kode.substring(3);
    }
  }

  render() {
    const { onSelect, query, searchResults, classes } = this.props;
    if (!searchResults) return null;
    if (searchResults.length <= 0) return null;
    return (
      <div
        //className="resultatliste" className={abc("resultatliste", isSearching)}
        //className="resultatliste"
        className="resultatliste mobile_active"
      >
        {searchResults.map(item => {
          const navn = item.title;
          return (
            <React.Fragment key={item.url}>
              <ListItem
                button={true}
                className={classes.listitem}
                onMouseDown={() => onSelect(item)}
                onKeyDown={e => {
                  if (e.keyCode === 13) {
                    onSelect(item);
                  }
                }}
                key={item.url}
              >
                <Bildeavatar url={item.url} />
                <ListItemText classes={{ primary: classes.text }}>
                  {ResultatListe.highlightMatch(navn, query, classes)}
                </ListItemText>

                <div className="itemtext">
                  {ResultatListe.highlightMatch(
                    this.filtrer(item.kode),
                    query,
                    classes
                  )}
                </div>
              </ListItem>
              <Divider className="inset" />
            </React.Fragment>
          );
        })}
      </div>
    );
  }

  static highlightMatch(text, query, classes) {
    if (!query) return text;
    const q = query.toLowerCase().split(" ")[0];
    const offset = text.toLowerCase().indexOf(q);
    if (offset < 0) return text;

    const end = offset + q.length;
    return (
      <span>
        {text.substring(0, offset)}
        <span className="textmatch">{text.substring(offset, end)}</span>
        {text.substring(end, text.length)}
      </span>
    );
  }
}

export default withStyles(styles)(ResultatListe);
