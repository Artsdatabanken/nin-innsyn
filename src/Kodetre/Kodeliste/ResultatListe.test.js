import ResultatListe from './ResultatListe'

test('Highlight', () => {
  expect(ResultatListe.highlightMatch('Karm√∏y', 'ka k')).toMatchSnapshot()
})
