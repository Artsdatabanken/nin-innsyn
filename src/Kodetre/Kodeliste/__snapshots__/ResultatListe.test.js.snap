// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlight 1`] = `
<React.Fragment>
   
  <span>
    
  </span>
  <span>
    Ka
  </span>
  <span>
    rm√∏y
  </span>
</React.Fragment>
`;
